client/reference-data.ts

```diff
@@ -0,0 +1,55 @@
+interface ReferenceDataMeta {
+  type: string;
+  hash: string;
+}
+
+type ReferenceDataType<T> = T extends Array<infer R> ? R : T;
+
+interface ReferenceData<T = unknown[]> extends ReferenceDataMeta {
+  value: Array<ReferenceDataType<T>>;
+}
+
+export class ReferenceDataStore {
+  store: LocalForage;
+
+  constructor(store: LocalForage) {
+    this.store = store;
+  }
+
+  async emptyStore() {
+    await this.store.clear();
+  }
+
+  async updateStore<T>(referenceData: ReferenceData[]) {
+    await this.store.clear();
+    for (const data of referenceData) {
+      await this.store.setItem(data.type, data);
+    }
+  }
+
+  async getReferenceDataByType<T>(
+    typeKey: string
+  ): Promise<ReferenceData<T> | null> {
```

Comment:
could we force the return to be `undefined` instead of `null` if there is no data to return? `null` will mean we will have to define types later on to account for null rather than using the optional modifier